### PR TITLE
Extract find bar and jump to definition from devtools to separate addons

### PR DESCRIPTION
**Pull request by TheColaber**
Related to https://github.com/ScratchAddons/ScratchAddons/issues/3700

### Changes

- New addon: find-bar
- New addon: jump-to-definition
- Removed help popup
- Dynamic Enable and disable for both new addon
- Search for sounds in find bar
- Action when find-bar is enabled but jump-to-definition is not.
- Transition code
- Bug fixes for find-bar.

### Devtools extension
I spoke to WL and he said that we will do one of two things.
1. Remove more and more features from the extension so that users can transition to SA
2. Bundle up this addon and the devtools addon together

Feel free to review if I've missed anything.

### Tests
I need people to test this.

### CODE REVIEW

**Comment by Samq64** [comment]
The only problem I've noticed so far is that dark mode styles are broken. Other than that it looks pretty good.

image

**Comment by Samq64** [comment]
Also, the help button doesn't work anymore.

**Comment by TheColaber** [comment]
> Also, the help button doesn't work anymore.

Yeah, I'm going to probably remove it since the addon is separate from devtools now.

**Comment by Weredime** [comment]
the folder `blockly` should be put into `libraries` because multiple addons use it

**Comment by Samq64** [comment]
Is the find label next to the textbox necessary? The textbox placeholder already says that.

File: addons-l10n/en/editor-devtools.json
Comment (Samq64) [inline addons-l10n/en/editor-devtools.json:8]: Wasn't this removed from devtools a while ago?

File: addons-l10n/en/find-bar.json
Comment (Samq64) [inline addons-l10n/en/find-bar.json:5]: There are also unused now?

**Comment by TheColaber** [comment]
> Is the find label next to the textbox necessary? The textbox placeholder already says that.

That can be changed in a later pr, this pr should keep everything relative the same as before.

**Comment by OregSamSas** [comment]
~~I think it doesn't work if language is set to a language other than English. I don't know if it's because it needs translations now, If I remember right `devtools` didn't need them, and it worked fine for me even in Hungarian.~~
(My Chrome is in Hungarian, and because loading extensions in are not just temporarily in Chrome, I used that one and found this, maybe luckily)

> ---
**Ohh, UPDATE:
Only for Hungarian doesn't it work.**
I had a look at on what about other languages, and the result was it was fine for them. I don't what's going on with HU.
> ---

 I changed the language (at first I didn't need to, because it was automatically in Hungarian), then middle clicked one of the message hat blocks, and this was the result:
image

I found out then that, there was only one event listed in the dropdown whose name was _"üzenet érkezésekor"_ which is the translation of "when I receive ..."
image

**Comment by OregSamSas** [comment]
And it's not that easy. 
It works if I middle click on a stack event block, but middle clicking jumps to the "üzenet érkezésekor" event, but doesn't give any search results for it.

<html>
<details>
<summary>
screenshots
</summary>
<p><code>#1</code>:<br>

<br>
<p><code>#2</code>:<br>

</details>
<details>
<summary>
Also  the &quot;event <i>&uuml;zenet &eacute;rkez&eacute;sekor</i>&quot; always shows up in the dropdown list
</summary>

</details>
</html>

**Comment by TheColaber** [comment]
@OregSamSas Fixed! This bug existed before this PR was created. Thanks for reporting.

**Comment by Samq64** [comment]
I have a few suggestions:

* It would look better if the menu faded or had another animation when opening and closing it.
* I don't like the colour of highlighted items in the costumes and sounds tabs.
* When using the keyboard to scroll through items, the selected item can go offscreen. It should scroll the menu.
* When the find bar is enabled, but jump to block definition isn't, middle-clicking custom blocks should open it in the find bar like variables.

**Comment by Secret-chest** [comment]
> It would look better if the menu faded or had another animation when opening and closing it.

It should roll down.

**Comment by mxmou** [comment]
> * When the find bar is enabled, but jump to block definition isn't, middle-clicking custom blocks should open it in the find bar like variables.

This was already suggested here: https://github.com/ScratchAddons/ScratchAddons/issues/3700#issuecomment-1188246800

I also think this is something this PR should change.

> * It would look better if the menu faded or had another animation when opening and closing it.
> * I don't like the colour of highlighted items in the costumes and sounds tabs. Maybe something like Scratch blue in the costume editor, and pink in the sound editor would be better.
> * When using the keyboard to scroll through items, the selected item can go offscreen. It should scroll the menu.

Could you open issues for these?

> > It would look better if the menu faded or had another animation when opening and closing it.
> 
> It should roll down.

That would look weird because the popup opens behind the input, not under it.

**Comment by TheColaber** [comment]
I implemented this table:
(thx lisa)
|   | **Middle-clicking non-definitions** | **Middle-clicking definitions**
| --- | --- | --- |
| **Find bar only** | Open in find bar | Open in find bar |
| **Jump to definition only** | Jump to definition | Do nothing |
| **Both** | Jump to definition | Open in find bar |

by polluting blockly which means this addon needs even more testing.

**Comment by Weredime** [comment]
@TheColaber Since both `jump-to-def` and `find-bar` uses the `blockly` folder, why not put it in `libraries/cs`?

**Comment by TheColaber** [comment]
No need to do so, it's fine for an addon to import other addon code like we do for the pause button and debugger.

**Review by Weredime — APPROVED** [review]
Tested and it works

**Comment by OregSamSas** [comment]
I have a suggestion, what I tried to make, but I just wasn't enough smart to succeed. I found that the find bar stays there on the forth tab too (if the `variable-manager` is enabled), so if we could hide its own search bar and use `find-bar`'s one to jump to specific variables, this would be awesome. Because Find bar can be accessed by Ctrl+F, and it looks more good.

Below it can be checked, what I've managed to do. I couldn't complete it (the main part was very hard), because it would have needed the communication between two addon, and I couldn't use that one properly.

The special features that needs to be added that on clicking an item in the dropdown would jump to it's value input field, and on shift click it would jump into the name field, and focus on it automatically.

Maybe I should make an issue for this, but I feel it's just too late for that. I rather finish this comment.
<html>
<details>
<summary>concept</summary>

</html>

**Comment by mxmou** [comment]
@OregSamSas I don't think a separate addon should be required to search the list of variables.

**Comment by OregSamSas** [comment]
Ok then just hide the find bar on the variable manager tab.

**Comment by TheColaber** [comment]
I think the best thing to do would be to leave the list of items in the find bar empty.

**Comment by mxmou** [comment]
> I think the best thing to do would be to leave the list of items in the find bar empty.

Why does it need to be visible if you can't use it to find anything?

**Comment by OregSamSas** [comment]
> @OregSamSas I don't think a separate addon should be required to search the list of variables.

It wouldn't be required, it would add just some cool features for the searching on the `variable-manager` addon, IF both addon were enabled.
Reasons: it doesn't look good that you have multiple ways to search for an item in the variable+lists list, with a find-bar which technically wouldn't work (because it would just list nothing, according to the current state of this addon declared by Colaber⁽¹⁾), and a search bar, which doesn't jump to searched items, but hides the items which didn't match with the key. I'm not totally satisfied with that, I would prefer the find bar instead.

So basically the current search bar for variable-manager would stay the same in my suggestion, just a new (new, only for that tab) find-bar would be added to options. (If  searching for variables on the 4th tab was working, there should be a setting called: "Replace variable manager's built-in search bar", which would then hide that search bar, to use the find-bar, which has/would have more conveniences.)

If there's anything I'm wrong about, just let me know, because I could imagine, I've misunderstood your response.
And so sorry for not speaking understandably, it does really annoy me too, just I can't change it sadly :( .

⁽¹⁾ Here's the quote:
> I think the best thing to do would be to leave the list of items in the find bar empty.

**Comment by mxmou** [comment]
> > Why does it need to be visible if you can't use it to find anything?
> 
> Following that logic, shouldn't Scratch hide the stop button icon if the project isn't running? I think it would be distracting if the find bar was hidden each time you switch to the variables tab.

It could also be disabled, but I don't think the reason why it's disabled would be obvious.

**Comment by apple502j** [comment]
> Following that logic, shouldn't Scratch hide the stop button icon if the project isn't running?

FYI, stop button isn't just for stopping threads.

> I think it would be distracting if the find bar was hidden each time you switch to the variables tab.

I think it's fine.

**Comment by BroJac5246** [comment]
> I think it would be distracting if the find bar was hidden each time you switch to the variables tab.

I personally disagree, although that's just me.


>  Following that logic, shouldn't Scratch hide the stop button icon if the project isn't running?

That should be an addon [Goes to suggest it] ;)

**Comment by WorldLanguages** [comment]
Note that we need griffpatch approval before merging this

**Comment by Weredime** [comment]
> Note that we need griffpatch approval before merging this

@griffpatch

**Comment by griffpatch** [comment]
Sure thing... Make it so.

File: addons/find-bar/addon.json
Comment (apple502j) [inline addons/find-bar/addon.json:3]:
```suggestion
  "description": "Adds a find bar to find and jump to scripts, costumes and sounds next to the sounds tab. Use Ctrl + Left and Ctrl + Right in the code area to navigate to previous or next visited position after using the find bar.",
```

File: addons/jump-to-def/addon.json
Comment (apple502j) [inline addons/jump-to-def/addon.json:3]:
```suggestion
  "description": "Allows you to jump to a custom block's definition using the middle mouse button or Shift + Click on the block.",
```

**Review by apple502j — CHANGES_REQUESTED** [review]

**Comment by apple502j** [comment]
Migration code is broken (version setter removed, see #4916 and also some code is messed up)

**Review by apple502j — APPROVED** [review]

**Comment by apple502j** [comment]
:shipit:

File: addons/editor-devtools/DevTools.js
Comment (WorldLanguages) [inline addons/editor-devtools/DevTools.js:2215]: I think you removed this without checking if other files reference this id.

This id should have been kept, it's referenced on Tab.js

**Pull request merged**
